lib/spack/spack/binary_distribution.py

```diff
@@ -239,35 +255,89 @@ def sign_tarball(key, force, specfile_path):
     Gpg.sign(key, specfile_path, '%s.asc' % specfile_path)
 
 
-def generate_index(outdir, indexfile_path):
-    f = open(indexfile_path, 'w')
+def generate_html_index(path_list, output_path):
+    f = open(output_path, 'w')
     header = """<html>\n
 <head>\n</head>\n
 <list>\n"""
     footer = "</list>\n</html>\n"
-    paths = os.listdir(outdir + '/build_cache')
     f.write(header)
-    for path in paths:
+    for path in path_list:
         rel = os.path.basename(path)
         f.write('<li><a href="%s"> %s</a>\n' % (rel, rel))
     f.write(footer)
     f.close()
 
 
+def generate_json_index(path_list, output_path):
+    """Read all the .spec.yaml files generated by build_tarball and
+       create an index from them."""
+    index_object = {}
+
+    for path in path_list:
+        m = _meta_yaml_regex.search(path)
+
+        if not m:
+            # Expecting that some things won't match because they're not a spec
+            # yaml file (like the index files written on a previous pass, or
+            # the directories containing tarballs).  Just skip those.
+            continue
+
+        pkg_name = m.group(1)
+        pkg_version = m.group(2)
+        pkg_dag_hash = m.group(3)
+
+        with open(path, 'r') as yaml_in:
+            yaml_str = yaml_in.read()
+            yaml_obj = yaml.load(yaml_str)
+
+            entry = {
+                'name': pkg_name,
+                'version': pkg_version
+            }
+
+            if 'full_hash' in yaml_obj:
+                entry['full_hash'] = yaml_obj['full_hash']
+
+            index_object[pkg_dag_hash] = entry
+
+    validate(index_object, schema)
+
+    with open(output_path, 'w') as f:
+        f.write(json.dumps(index_object))
+
+
+def generate_package_index(build_cache_dir):
+    yaml_list = os.listdir(build_cache_dir)
+    path_list = [os.path.join(build_cache_dir, l) for l in yaml_list]
+
+    index_html_path = os.path.join(build_cache_dir, 'index.html')
+    index_json_path = os.path.join(build_cache_dir, 'index.json')
+
+    for p in [index_html_path, index_json_path]:
+        if os.path.exists(p):
+            os.remove(p)
+
+    generate_html_index(path_list, index_html_path)
+    generate_json_index(path_list, index_json_path)
```

Comment:
instead of removing then generating in-place, can you generate to a `.index.json.tmp` file, and then move that on top of `index.json`?  That way the change is atomic and you don't risk failing with missing or partial indices.  The local name/suffix (as opposed to tmp or something) guarantees it's on the same filesystem, so a move will be atomic.

Reply:
Sounds good, I'll do it.